src/app/data-model/IteratableList.h

```diff
@@ -0,0 +1,172 @@
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include "Encode.h"
+#include <iterator>
+#include <lib/core/CHIPCore.h>
+#include <lib/core/CHIPSafeCasts.h>
+#include <lib/core/CHIPTLV.h>
+#include <lib/core/Optional.h>
+#include <lib/support/CodeUtils.h>
+
+namespace chip {
+namespace app {
+namespace DataModel {
+
+/*
+ * @brief
+ *
+ * This class provides an iteratable decoder of list items within TLV payloads
+ * such that no memory has to be provided ahead of time to store the entirety of the decoded
+ * list contents.
+ *
+ */
+template <typename T>
+class IteratableList
+{
+public:
+    /*
+     * @brief
+     *
+     * This calls stores a TLV reader positioned on the list this class is to manage.
+     *
+     * Specifically, the passed-in TLV reader should be pointing into the list just after
+     * having called `OpenContainer` on the list element.
+     */
+    void SetReader(const TLV::TLVReader & reader) { mReader = reader; }
+
+    size_t size()
+    {
+        CHIP_ERROR err;
+        TLV::TLVReader reader;
+        reader.Init(mReader);
+        size_t count = 0;
+
+        while ((err = reader.Next()) == CHIP_NO_ERROR)
+        {
+            count++;
+        }
+
+        return count;
+    }
+
+    class Iterator
+    {
+    public:
+        using iterator_category = std::input_iterator_tag;
+
+        /*
+         * Initialize the iterator with a reference to a reader.
+         *
+         * This reader should be pointing into the list just after
+         * having called `OpenContainer` on the list element.
+         */
+        Iterator(const TLV::TLVReader & reader)
+        {
+            mStatus = CHIP_NO_ERROR;
+            mReader.Init(reader);
+        }
+
+        /*
+         * Increments the iterator to point to the next list element
+         * if a valid one exists, and decodes the list element into
+         * the internal value storage.
+         *
+         * If an element does exist and was successfully decoded, this
+         * shall return true.
+         *
+         * Otherwise, if any error was encountered OR the end of the list
+         * was reached, this shall return false.
+         *
+         */
+        bool Next()
+        {
+            if (mStatus != CHIP_NO_ERROR)
+            {
+                return false;
+            }
+
+            mStatus = mReader.Next();
+
+            if (mStatus == CHIP_NO_ERROR)
```

Comment:
I wonder about:

```cpp
bool Next() 
{

  if (mStatus == CHIP_NO_ERROR) 
  {
    mStatus = mReader.Next();
  }

  if (mStatus == CHIP_NO_ERROR) 
  {
    mStatus = Decode(mReader, mValue); 
  }

  return (mStatus == CHIP_NO_ERROR);
}
```

because it feels shorter and makes it obvious the 'this returns true if no error, false if error